File: parsers/valvechveiculos.py
# parsers/valvechveiculos.py
from bs4 import BeautifulSoup
from parsers.base_parser import BaseParser
from config import REQUEST_TIMEOUT

class ValvechVeiculosParser(BaseParser):
    def search(self, query):
        url = f"{self.base_url}/_Geral/getEstoque"
        
        results = []
        page = 1
        query_words = query.lower().split()
        
        while page <= 10:  # Limita a busca a no máximo 10 páginas
            try:
                # O site espera o número da página no POST
                headers = {"X-Requested-With": "XMLHttpRequest"}
                response = self.session.post(
                    url, 
                    data={"pagina": str(page)}, 
                    headers=headers,
                    timeout=REQUEST_TIMEOUT
                )
                response.raise_for_status()
                
                html = response.text
                try:
                    data = response.json()
                    html = data.get("html", "")
                except Exception:
                    pass
                
                soup = BeautifulSoup(html, "html.parser")
                cards = soup.find_all("div", class_="cerca")
                
                if not cards:
                    break
                    
                for card in cards:
                    try:
                        # 1. URL do anúncio
                        a_tag = card.find("a", href=True)
                        if not a_tag:
                            continue
                        
                        link = a_tag["href"]
                        if link.startswith("//"):
                            link = f"https:{link}"
                        elif not link.startswith("http"):
                            link = f"{self.base_url.rstrip('/')}/{link.lstrip('/')}"
                            
                        # Limpa possíveis barras duplas na URL construída
                        link = link.replace("com.br//", "com.br/")
                        
                        # 2. Título
                        title_span = card.find("span", class_=lambda c: c and 'l' in c.split())
                        if not title_span:
                            title_span = card.find("span", class_="l")
                            
                        title = title_span.get_text(strip=True) if title_span else "Veículo"
                        title = " ".join(title.split())
                        
                        # Filtro de busca local
                        if not all(word in title.lower() for word in query_words):
                            continue
                            
                        # 3. Imagem
                        img_tag = card.find("img")
                        image_url = img_tag.get("src") if img_tag else None
                        if image_url and image_url.startswith("//"):
                            image_url = f"https:{image_url}"
                        elif image_url and not image_url.startswith("http"):
                            image_url = f"{self.base_url.rstrip('/')}/{image_url.lstrip('/')}"
                            
                        # 4. Preço
                        price_span = card.find("span", class_=lambda c: c and 'r' in c.split())
                        if not price_span:
                            price_span = card.find("span", class_="r")
                            
                        price = price_span.get_text(strip=True) if price_span else "Sob consulta"
                        price = " ".join(price.split())
                        
                        if price == "R$ 0,00" or price == "R$":
                            price = "Sob consulta"
                            
                        results.append({
                            "title": title,
                            "price": price,
                            "image_url": image_url,
                            "url": link,
                            "dealer_name": self.name
                        })
                    except Exception as e:
                        print(f"Erro ao processar veículo na {self.name}: {e}")
                        continue
                
                # Se a página retornou menos carros que o tamanho padrão de página (24), terminamos
                if len(cards) < 24:
                    break
                    
                page += 1
                
            except Exception as e:
                print(f"Erro ao pesquisar em {self.name} (Pág. {page}): {e}")
                break
                
        return results

File: api/_scrapers/valvechveiculos.js
// api/parsers/valvechveiculos.js
// Valvech Veículos — POST /_Geral/getEstoque paginado — filtro local

const cheerio = require("cheerio");
const NAME     = "Valvech Veículos";
const BASE_URL = "https://valvechveiculos.com.br";

async function search(query, fetchHtml) {
  const queryWords = query.toLowerCase().split(/\s+/);
  const results    = [];

  for (let page = 1; page <= 5; page++) {
    try {
      // POST com paginação
      const raw = await fetch(`${BASE_URL}/_Geral/getEstoque`, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          "X-Requested-With": "XMLHttpRequest",
          "User-Agent": "Mozilla/5.0",
        },
        body: `pagina=${page}`,
      });
      let html = await raw.text();
      try { const j = JSON.parse(html); html = j.html || html; } catch (_) {}

      const $     = cheerio.load(html);
      const cards = $(".cerca");
      if (!cards.length) break;

      cards.each((_, card) => {
        try {
          const el = $(card);

          let link = el.find("a[href]").first().attr("href") || null;
          if (link?.startsWith("//")) link = `https:${link}`;
          else if (link && !link.startsWith("http")) link = `${BASE_URL}/${link.replace(/^\//, "")}`;
          if (!link) return;

          let title = el.find("span.l").text().trim().replace(/\s+/g, " ") || "Veículo";
          if (!queryWords.every(w => title.toLowerCase().includes(w))) return;

          let image_url = el.find("img").first().attr("src") || null;
          if (image_url?.startsWith("//")) image_url = `https:${image_url}`;
          else if (image_url && !image_url.startsWith("http")) image_url = `${BASE_URL}/${image_url.replace(/^\//, "")}`;

          let price = el.find("span.r").text().trim().replace(/\s+/g, " ") || "Sob consulta";
          if (price === "R$ 0,00" || price === "R$") price = "Sob consulta";

          
      // --- Auto-Extraction of Year and KM ---
      let extYear = null;
      let extKm = null;
      try {
        const fullText = el.text().replace(/\s+/g, " ");
        // Year: like 2019/2020 or 2019
        const yearMatch = fullText.match(/\b(20\d{2}|19\d{2})(?:\/[12]0\d{2})?\b/);
        if (yearMatch) extYear = yearMatch[0];
        
        // KM: like 45.000 km, 120 mil km, 45000km
        let kmMatch = fullText.match(/\bkm[:\-\s]*(\d{1,3}(?:\.\d{3})+|\d+)\b/i);
        if (!kmMatch) kmMatch = fullText.match(/\b(\d{1,3}(?:\.\d{3})+|\d+)\s*(?:km|kms|mil\s*km)\b/i);
        if (kmMatch) {
          let parsedKm = parseInt(kmMatch[1].replace(/\./g, ""), 10);
          if (extYear && parsedKm === parseInt(extYear)) kmMatch = null;
          if (kmMatch && !isNaN(parsedKm)) extKm = new Intl.NumberFormat("pt-BR").format(parsedKm) + " km";
        }
      } catch (e) {}

      results.push({ title, price, image_url, url: link, dealer_name: NAME, year: extYear, km: extKm });
        } catch (_) {}
      });

      if (cards.length < 24) break;
    } catch (_) { break; }
  }
  return results;
}
module.exports = { search, name: NAME };
